documentation/dataplane/dash-bmv2-data-plane-app.md

```diff
@@ -0,0 +1,383 @@
+
+**Note:** For simplicity and concept verification, worker threads may directly call DASH SAI to offload flow to BMv2. The concern is that DASH SAI blocking API can block packet processing of workers.
+
+## 6. Detailed design
+
+Referring to the below figure from [HA API HLD](https://github.com/sonic-net/DASH/blob/main/documentation/high-avail/ha-api-hld.md), it greatly outlines the whole packet flow in data plane for both standalone and HA context.
+
+![packet flow in data plane](https://github.com/sonic-net/DASH/blob/main/documentation/high-avail/images/ha-bm-packet-flow.svg)
+
+From the perspective of DPAPP, its core task in slow path is to create flow in flow table in case of flow miss in Flow (Conntrack) lookup stage, and then maintain flow state. The sub sections will depict these functions around flow object.
+
+
+### 6.1. DASH metadata
+DASH metadata records the packet processing result of DASH pipeline. It can have the following info to help flow creation/removal, etc operations:
+- flow key entry - eni_mac, vnet_id, 5-tuples
+- common flow data - state, direction, eni, flow actions, metering class, policy ID
+- flow overlay rewrite data 
+- flow underlay encapsulation data 
+- flow tunnel encapsulation data 
+
+When DASH pipeline requests DPAPP for flow creation, it encapsulates DASH metadata in an ethernet frame with EtherType DASH_METADATA and appends the original customer packet. The packet sent to DPAPP is like:
+
+```
+    Ethernet HEADER | DASH metadata | customer packet
+```
+
+The number of DASH_METADATA is 0x876D, which reuses the number of EtherType SECURE_DATA (vpp/src/vnet/ethernet/types.def at master · FDio/vpp · GitHub).
+DASH metadata is encoded in protocol DASH_METADATA, whose message format is defined as the below figure:
+
+```
+    0                   1                   2                   3   
+    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |      Source   | Type  |Subtype|           Length              |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |                           Flow key entry                      |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |                           Flow data                           :
+   :                                                               |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+   Source: Where is packet from, dash_packet_source_t
+   Type: Packet type under certain source
+   Subtype: Packet subtype under certain type
+   Length: Total length of DASH header
+   Flow key entry: eni_mac, 5-tuples
+   Flow data: As next figure
+```
+
+```
+   0                   1                   2                   3   
+   0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 
```

Comment:
we can use the mermaid graph to simplify this graph now, which is just available on github.

Reply:
mermaid packet diagram is still in beta, not support `varbit` representation. I'll prefer to it after it has this feature.

Reply:
i see. sounds good!